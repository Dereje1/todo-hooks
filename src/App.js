import './App.css';
import React, { Fragment } from 'react';
import taskListStub from './taskliststub';
/* MUI */
import Box from '@mui/material/Box';
import Collapse from '@mui/material/Collapse';
import IconButton from '@mui/material/IconButton';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import KeyboardArrowUpIcon from '@mui/icons-material/KeyboardArrowUp';

function Row(props) {
  const { row } = props;
  const [open, setOpen] = React.useState(false);
  return (
    <Fragment>
      <TableRow sx={{ '& > *': { borderBottom: 'unset' } }}>
        <TableCell>
          <IconButton
            aria-label="expand row"
            size="small"
            onClick={() => setOpen(!open)}
          >
            {open ? <KeyboardArrowUpIcon /> : <KeyboardArrowDownIcon />}
          </IconButton>
        </TableCell>
        <TableCell component="th" scope="row">
          {row.name}
        </TableCell>
        {/* <TableCell align="right">{row.calories}</TableCell>
        <TableCell align="right">{row.fat}</TableCell>
        <TableCell align="right">{row.carbs}</TableCell>
        <TableCell align="right">{row.protein}</TableCell> */}
      </TableRow>
      <TableRow>
        <TableCell style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={6}>
          <Collapse in={open} timeout="auto" unmountOnExit>
            <Box sx={{ margin: 1 }}>
              {row.description}
            </Box>
          </Collapse>
        </TableCell>
      </TableRow>
    </Fragment>
  );
}

const App = () => (
  <TableContainer component={Paper}>
    <Table aria-label="collapsible table">
      <TableBody>
        {taskListStub.map((row) => (
          <Row key={row.name} row={row} />
        ))}
      </TableBody>
    </Table>
  </TableContainer>
)

export default App;
