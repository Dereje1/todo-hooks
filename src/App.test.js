import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import App, { Row } from './App';

describe('App',()=>{
  test('renders table', () => {
    const wrapper = shallow(<App />)
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test('Will handle task completion',()=>{
    const wrapper = shallow(<App />);
    let row = wrapper.find('Row').at(1)
    expect(row.props().row.completed).toBe(false);
    row.props().onCompleteTask({target:{id: '2'}});
    row = wrapper.find('Row').at(1)
    expect(row.props().row.completed).toBe(true);
  })
})

describe('rows', () => {
  let rowProps;
  beforeEach(() => {
    rowProps = {
      row: {
        id: "1",
        name: "Task 1",
        completed: false,
        description: "This is task # 1"
      },
      onCompleteTask: jest.fn()
    }
  })
  afterEach(() => {
    rowProps = null;
  })
  test('renders row', () => {
    const wrapper = shallow(<Row {...rowProps}/>)
    expect(toJson(wrapper)).toMatchSnapshot();
  })
  test('expands a row', () => {
    const wrapper = shallow(<Row {...rowProps}/>)
    let collapseableRow = wrapper.find('ForwardRef(Collapse)');
    expect(collapseableRow.props().in).toBe(false);
    const collapseButton = wrapper.find({"aria-label":"expand row"})
    collapseButton.props().onClick();
    collapseableRow = wrapper.find('ForwardRef(Collapse)');
    expect(collapseableRow.props().in).toBe(true);
  })
  test('Will toggle to complete a task', ()=>{
    const wrapper = shallow(<Row {...rowProps}/>)
    const completedToggle = wrapper.find('ForwardRef(Checkbox)')
    completedToggle.props().onChange();
    expect(rowProps.onCompleteTask).toHaveBeenCalled()
  })
})

