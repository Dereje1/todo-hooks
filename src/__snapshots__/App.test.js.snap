// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders table 1`] = `
<Fragment>
  <div
    id="button-group-container"
    style={
      Object {
        "display": "flex",
        "justifyContent": "center",
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    <ForwardRef(ButtonGroup)
      aria-label="outlined primary button group"
      sx={
        Object {
          "padding": 1,
        }
      }
      variant="contained"
    >
      <ForwardRef(Tooltip)
        title="Add a task"
      >
        <ForwardRef(IconButton)
          id="add-task"
          onClick={[Function]}
        >
          <Memo(ForwardRef(AddIcon)) />
        </ForwardRef(IconButton)>
      </ForwardRef(Tooltip)>
      <ForwardRef(Tooltip)
        title="Completed tasks"
      >
        <ForwardRef(IconButton)
          id="completed-tasks"
          onClick={[Function]}
        >
          <Memo(ForwardRef(CheckIcon))
            color="primary"
          />
        </ForwardRef(IconButton)>
      </ForwardRef(Tooltip)>
      <ForwardRef(Tooltip)
        title="Active tasks"
      >
        <ForwardRef(IconButton)
          id="active-tasks"
          onClick={[Function]}
        >
          <Memo(ForwardRef(NotificationImportantIcon))
            color="error"
          />
        </ForwardRef(IconButton)>
      </ForwardRef(Tooltip)>
      <ForwardRef(Tooltip)
        title="All tasks"
      >
        <ForwardRef(IconButton)
          id="all-tasks"
          onClick={[Function]}
        >
          <Memo(ForwardRef(AllInclusiveIcon))
            color="success"
          />
        </ForwardRef(IconButton)>
      </ForwardRef(Tooltip)>
    </ForwardRef(ButtonGroup)>
  </div>
  <ForwardRef(TableContainer)
    component={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "propTypes": Object {
          "children": [Function],
          "className": [Function],
          "classes": [Function],
          "component": [Function],
          "elevation": [Function],
          "square": [Function],
          "sx": [Function],
          "variant": [Function],
        },
        "render": [Function],
      }
    }
    sx={
      Object {
        "margin": "0 auto",
        "maxWidth": 700,
      }
    }
  >
    <ForwardRef(Table)
      aria-label="collapsible table"
    >
      <ForwardRef(TableHead)>
        <ForwardRef(TableRow)>
          <ForwardRef(TableCell) />
          <ForwardRef(TableCell) />
          <ForwardRef(TableCell)>
            Completed
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)>
            Delete
          </ForwardRef(TableCell)>
        </ForwardRef(TableRow)>
      </ForwardRef(TableHead)>
      <ForwardRef(TableBody)>
        <Row
          key="1"
          onCompleteTask={[Function]}
          onDeleteTask={[Function]}
          row={
            Object {
              "completed": false,
              "description": "This is task # 1",
              "id": "1",
              "name": "Task 1",
            }
          }
        />
        <Row
          key="2"
          onCompleteTask={[Function]}
          onDeleteTask={[Function]}
          row={
            Object {
              "completed": false,
              "description": "This is task # 2",
              "id": "2",
              "name": "Task 2",
            }
          }
        />
        <Row
          key="3"
          onCompleteTask={[Function]}
          onDeleteTask={[Function]}
          row={
            Object {
              "completed": false,
              "description": "This is task # 3",
              "id": "3",
              "name": "Task 3",
            }
          }
        />
      </ForwardRef(TableBody)>
    </ForwardRef(Table)>
  </ForwardRef(TableContainer)>
</Fragment>
`;

exports[`rows renders row 1`] = `
<Fragment>
  <ForwardRef(TableRow)
    sx={
      Object {
        "& > *": Object {
          "borderBottom": "unset",
        },
      }
    }
  >
    <ForwardRef(TableCell)>
      <ForwardRef(IconButton)
        aria-label="expand row"
        onClick={[Function]}
        size="small"
      >
        <Memo(ForwardRef(KeyboardArrowDownIcon)) />
      </ForwardRef(IconButton)>
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      component="th"
      scope="row"
    >
      Task 1
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
      padding="checkbox"
    >
      <ForwardRef(Checkbox)
        checked={false}
        color="primary"
        id="1"
        inputProps={
          Object {
            "aria-label": "mark complete",
          }
        }
        onChange={[MockFunction]}
      />
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
      padding="checkbox"
    >
      <ForwardRef(IconButton)
        id="delete-task"
        onClick={[Function]}
      >
        <Memo(ForwardRef(DeleteForeverIcon))
          color="error"
          sx={
            Object {
              "fontSize": "1.3em",
            }
          }
        />
      </ForwardRef(IconButton)>
    </ForwardRef(TableCell)>
  </ForwardRef(TableRow)>
  <ForwardRef(TableRow)>
    <ForwardRef(TableCell)
      colSpan={6}
      style={
        Object {
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <ForwardRef(Collapse)
        in={false}
        timeout="auto"
        unmountOnExit={true}
      >
        <ForwardRef(Box)
          sx={
            Object {
              "margin": 1,
            }
          }
        >
          This is task # 1
        </ForwardRef(Box)>
      </ForwardRef(Collapse)>
    </ForwardRef(TableCell)>
  </ForwardRef(TableRow)>
</Fragment>
`;
